refactor: Convert GenTxid to `std::variant`

marcofleon opened the pull request:
Part of the [type safety refactor](https://github.com/bitcoin/bitcoin/pull/32189).

This PR changes the GenTxid class to a variant, which holds both Txids and Wtxids. This provides compile-time type safety and eliminates the manual type check (bool m_is_wtxid). Variables that can be either a Txid or a Wtxid are now using the new GenTxid variant, instead of uint256.

marcofleon committed:
Implement GenTxid as a variant

Reimplements the GenTxid class as a variant for better type safety.
Also adds two temporary functions to the old GenTxid class that
convert to and from the new variant.

marcofleon committed:
Replace GenTxid with Txid/Wtxid overloads in `txmempool`

Co-authored-by: stickies-v <stickies-v@protonmail.com>

marcofleon committed:
Convert `CompareInvMempoolOrder` to GenTxidVariant

Now that we are storing `CTxMemPool::CompareDepthAndScore` parameters using
`std::variant` we have no portable zero-overhead way of accessing them,
so use `std::visit` and drop `bool wtxid` in-parameter.

Co-authored-by: stickies-v <stickies-v@protonmail.com>

marcofleon committed:
Convert `txdownloadman_impl` to GenTxidVariant

Convert all of `txdownloadman_impl` to the new variant except for
`GetRequestsToSend`, which will be easier to switch at the same
time as `txrequest`.

marcofleon committed:
Convert `txrequest` to GenTxidVariant

Switch all instances of GenTxid to the new variant
in `txrequest` and complete `txdownloadman_impl` by
converting `GetRequestsToSend`.

marcofleon committed:
Convert remaining instances of GenTxid to GenTxidVariant

marcofleon committed:
Remove old GenTxid class

marcofleon commented:
In `txrequest` the `ByPeer` and `ByTxHash` indexes still sort announcements by the underlying uint256 txhash only, as opposed to using the GenTxid comparators which sort by type and then hash. The patch below can be used to make sure that there are no accidental switches from uint256 comparisons to GenTxid comparisons.
```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 011e3cd928..1a47a39dfd 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -206,6 +206,12 @@ struct QueuedBlock {
     std::unique_ptr<PartiallyDownloadedBlock> partialBlock;
 };
 
+struct CompareGenTxid {
+    bool operator()(const GenTxid& a, const GenTxid& b) const {
+        return std::tuple(a.index(), a.ToUint256()) < std::tuple(b.index(), b.ToUint256());
+    }
+};
+
 /**
  * Data structure for an individual peer. This struct is not protected by
  * cs_main since it does not contain validation-critical data.
@@ -302,7 +308,7 @@ struct Peer {
          *  non-wtxid-relay peers, wtxid for wtxid-relay peers). We use the
          *  mempool to sort transactions in dependency order before relay, so
          *  this does not have to be sorted. */
-        std::set<GenTxid> m_tx_inventory_to_send GUARDED_BY(m_tx_inventory_mutex);
+        std::set<GenTxid, CompareGenTxid> m_tx_inventory_to_send GUARDED_BY(m_tx_inventory_mutex);
         /** Whether the peer has requested us to send our complete mempool. Only
          *  permitted if the peer has NetPermissionFlags::Mempool or we advertise
          *  NODE_BLOOM. See BIP35. */
@@ -856,7 +862,7 @@ private:
     std::shared_ptr<const CBlock> m_most_recent_block GUARDED_BY(m_most_recent_block_mutex);
     std::shared_ptr<const CBlockHeaderAndShortTxIDs> m_most_recent_compact_block GUARDED_BY(m_most_recent_block_mutex);
     uint256 m_most_recent_block_hash GUARDED_BY(m_most_recent_block_mutex);
-    std::unique_ptr<const std::map<GenTxid, CTransactionRef>> m_most_recent_block_txs GUARDED_BY(m_most_recent_block_mutex);
+    std::unique_ptr<const std::map<GenTxid, CTransactionRef, CompareGenTxid>> m_most_recent_block_txs GUARDED_BY(m_most_recent_block_mutex);
 
     // Data about the low-work headers synchronization, aggregated from all peers' HeadersSyncStates.
     /** Mutex guarding the other m_headers_presync_* variables. */
@@ -2027,7 +2033,7 @@ void PeerManagerImpl::NewPoWValidBlock(const CBlockIndex *pindex, const std::sha
         std::async(std::launch::deferred, [&] { return NetMsg::Make(NetMsgType::CMPCTBLOCK, *pcmpctblock); })};
 
     {
-        auto most_recent_block_txs = std::make_unique<std::map<GenTxid, CTransactionRef>>();
+        auto most_recent_block_txs = std::make_unique<std::map<GenTxid, CTransactionRef, CompareGenTxid>>();
         for (const auto& tx : pblock->vtx) {
             most_recent_block_txs->emplace(tx->GetHash(), tx);
             most_recent_block_txs->emplace(tx->GetWitnessHash(), tx);
diff --git a/src/util/transaction_identifier.h b/src/util/transaction_identifier.h
index 3cdeb664d4..6ab5d98263 100644
--- a/src/util/transaction_identifier.h
+++ b/src/util/transaction_identifier.h
@@ -88,13 +88,12 @@ public:
         return std::visit([](const auto& id) -> const uint256& { return id.ToUint256(); }, *this);
     }
 
-    friend bool operator==(const GenTxid& a, const GenTxid& b) {
-        return a.index() == b.index() && a.ToUint256() == b.ToUint256();
-    }
-
-    friend bool operator<(const GenTxid& a, const GenTxid& b) {
-        return std::tuple(a.index(), a.ToUint256()) < std::tuple(b.index(), b.ToUint256());
-    }
+    friend bool operator==(const GenTxid& a, const GenTxid& b) = delete;
+    friend bool operator!=(const GenTxid& a, const GenTxid& b) = delete;
+    friend bool operator<(const GenTxid& a, const GenTxid& b) = delete;
+    friend bool operator<=(const GenTxid& a, const GenTxid& b) = delete;
+    friend bool operator>(const GenTxid& a, const GenTxid& b) = delete;
+    friend bool operator>=(const GenTxid& a, const GenTxid& b) = delete;
 };
 
 #endif // BITCOIN_UTIL_TRANSACTION_IDENTIFIER_H
 
 ```
 Then run `cmake --build build --target bitcoind`. To check all tests as well, comment out the unit and fuzz tests for `txrequest`, as these tests directly compare GenTxid objects.

dergoegge commented on src/test/fuzz/txdownloadman.cpp:
```diff
@@ -373,9 +376,12 @@ FUZZ_TARGET(txdownloadman_impl, .init = initialize)
-                GenTxid gtxid = fuzzed_data_provider.ConsumeBool() ?
-                                GenTxid::Txid(rand_tx->GetHash()) :
-                                GenTxid::Wtxid(rand_tx->GetWitnessHash());
+                GenTxid gtxid;
+                if (fuzzed_data_provider.ConsumeBool()) {
+                    rand_tx->GetHash();
+                } else {
+                    rand_tx->GetWitnessHash();
+                }
```
The return values for `GetHash` and `GetWitnessHash` need to be assigned to gtxid

dergoegge reviewed (commented):
Concept ACK

I'm not sure if the intermediate commits make things better here. Looking at the final diff seems simple enough.

stickies-v reviewed (commented):
Concept ACK, the new `GenTxid` allows us to better benefit from the type safety of `Txid` and `Wtxid` and this PR takes a clear and structured approach that seems straightforward to review.

Approach-wise, I think I have a slightly different view. I find code generally easier to understand when we use the underlying types {`Txid`, `Wtxid`} directly and as such would prefer functions to be overloaded with those types, pushing the handling of variants to the edge as much as possible. This also has the side benefit of gently nudging users of those functions to not using variants if they don't have to. I have left 2 such comments on the first commits, but I suspect it applies to later commits too (did not investigate that in detail yet). This is largely a style preference, and I don't want to block the PRs progress over it if other reviewers disagree.

theStack commented:
Concept ACK

marcofleon commented on src/util/transaction_identifier.h:
```diff
@@ -76,4 +78,23 @@ using Txid = transaction_identifier<false>;
+
+    friend bool operator<(const GenTxid& a, const GenTxid& b) {
+        return std::tuple(a.index(), a.ToUint256()) < std::tuple(b.index(), b.ToUint256());
+    }
```
done

marcofleon commented on src/txmempool.cpp:
```diff
@@ -893,7 +901,11 @@ CTransactionRef CTxMemPool::get(const uint256& hash) const
+    indexed_transaction_set::const_iterator i = std::visit(util::Overloaded{
+        [this](const Wtxid& wtxid) EXCLUSIVE_LOCKS_REQUIRED(cs) { return get_iter_from_wtxid(wtxid); },
+        [this](const Txid& txid) EXCLUSIVE_LOCKS_REQUIRED(cs) { return mapTx.find(txid); }
+    }, gtxid);
```
Thank you for the patch, it helped.

hodlinator reviewed (commented):
Concept ACK b740b6f9479a2351cd3b66a0f6e53848452b6bf8

hodlinator reviewed (commented):
Code review d5a6e7a0a40bb78c2fcadd2662f2d64b86c374b7

Thanks for incorporating most of my feedback!

Mostly focused on reviewing non-test code.

marcofleon commented on src/txrequest.cpp:
```diff
@@ -435,7 +426,7 @@ class TxRequestTracker::Impl {
             // The next best CANDIDATE_READY, if any, immediately precedes the REQUESTED or CANDIDATE_BEST
             // announcement in the ByTxHash index.
-            if (it_prev->m_txhash == it->m_txhash && it_prev->GetState() == State::CANDIDATE_READY) {
+            if (it_prev->m_txhash.ToUint256() == it->m_txhash.ToUint256() && it_prev->GetState() == State::CANDIDATE_READY) {
```
Yeah the inv should match the peer's wtxid relay setting. The announcements are converted to GenTxids in `ProcessMessage` and that's what's passed into `AddTxAnnouncement` and used throughout `txdownloadman_impl` and `txrequest`.

I would prefer to keep it all in this PR, as I don't want both the old GenTxid class and the new variant in master. But I agree it's a bit iffy to have to manually ensure that we didn't miss a place where a previous `uint256` comparison is now suddenly a `GenTxid` comparison. The patch I provide at the top should catch anywhere that I might've missed.

Yes, with this PR's change, we would have be intentional about when we're comparing GenTxid objects (type first, then hash) and when we're comparing only the underlying hash. Another option would be to make the patch an actual part of the code, deleting the GenTxid comparators and having custom comparators wherever GenTxids are involved. This felt like overkill to me though, and I think the point of this type safety refactor is to, from now on, differentiate between the transaction identifiers and the underlying transaction hash. In that case, it makes sense to me to have a standard way to compare GenTxids and then make the conversion to `uint256` explicit if comparison of the hash is needed.

marcofleon commented on src/net_processing.cpp:
```diff
@@ -5810,14 +5808,15 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                         vInvTx.pop_back();
-                        uint256 hash = *it;
-                        CInv inv(peer->m_wtxid_relay ? MSG_WTX : MSG_TX, hash);
+                        GenTxidVariant hash = *it;
```
We erase the iterator soon after but still use the hash, so making a copy here is correct.

marcofleon commented on src/primitives/transaction.h:
```diff
@@ -437,6 +437,19 @@ class GenTxid
+
+    static GenTxid FromVariant(const GenTxidVariant& variant) {
+        return GenTxid{
+            variant.index() == 1, // true if Wtxid, false if Txid
```
Done.

marcofleon committed:
refactor: make CTxMemPool::GetIter strongly typed

This allows adding a GetIter(const Wtxid&) overload in a next
commit, making it easier to visit this function from a variant.

Co-authored-by: stickies-v <stickies-v@protonmail.com>

stickies-v committed:
refactor: replace get_iter_from_wtxid with GetIter(const Wtxid&)

Overloading GetIter makes it easier to use std::visit patterns
from a GenTxid.

stickies-v committed:
move-only: make GetInfo a private CTxMemPool member

This allows it to be used by templated functions in a future commit.

stickies-v reviewed (commented):
I think the changes here can be simplified quite a bit by first replacing `get_iter_from_wtxid` with a `GetIter(const wtxid&)` overload, see my suggestions [here](https://github.com/bitcoin/bitcoin/pull/32631#discussion_r2163728201) and [here](https://github.com/bitcoin/bitcoin/pull/32631#discussion_r2163747520) (or full branch https://github.com/bitcoin/bitcoin/compare/master...stickies-v:bitcoin:2025-06/overload-getiter).

There are also quite a few clang-tidy violations that would be helpful to address here, left per-commit comments.

marcofleon committed:
scripted-diff: Replace GenTxidVariant with GenTxid

-BEGIN VERIFY SCRIPT-
sed -i 's/GenTxidVariant/GenTxid/g' $(git grep -l 'GenTxidVariant')
-END VERIFY SCRIPT-

stickies-v reviewed (commented):
Approach ACK 700819f87b019f854340886dc8d35e66e813b53f

I did a full code review and while the changes are numerous, type-safety or pre-existing usage of `GenTxid::{Txid,Wtxid}` makes many of them trivially obvious to be a refactor, which in my opinion makes this PR reviewable even without in-depth knowledge of the p2p code this PR touches.

marcofleon commented on src/util/transaction_identifier.h:
```diff
@@ -76,4 +81,25 @@ using Txid = transaction_identifier<false>;
 using Wtxid = transaction_identifier<true>;
 
+template <typename T>
+concept TransactionIdentifier = std::is_same_v<T, Txid> || std::is_same_v<T, Wtxid>;
```
Done. I didn't really know the best name for this initially. I like `TxidOrWtxid` a bit better.

marcofleon commented on src/net_processing.cpp:
```diff
@@ -2437,7 +2439,9 @@ void PeerManagerImpl::ProcessGetData(CNode& pfrom, Peer& peer, const std::atomic
-        CTransactionRef tx = FindTxForGetData(*tx_relay, ToGenTxid(inv));
+        CTransactionRef tx{inv.IsMsgWtx() ?
+                               FindTxForGetData(*tx_relay, Wtxid::FromUint256(inv.hash)) :
+                               FindTxForGetData(*tx_relay, Txid::FromUint256(inv.hash))};
```
Took this suggestion 🤝

marcofleon commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -122,33 +123,34 @@ void TxDownloadManagerImpl::BlockDisconnected()
-        if (m_orphanage.HaveTx(Wtxid::FromUint256(hash))) return true;
+bool TxDownloadManagerImpl::AlreadyHaveTx(const GenTxidVariant& gtxid, bool include_reconsiderable)
+{
+    if (std::visit(util::Overloaded{
```
I agree that there isn't a need for branching on wtxid/txid and that having the comment there is enough. Just to be sure, @glozow if you prefer to keep the branching, let me know and I can revert.

marcofleon commented on src/node/interfaces.cpp:
```diff
@@ -673,7 +673,7 @@ class ChainImpl : public Chain
         if (!m_node.mempool) return false;
         LOCK(m_node.mempool->cs);
-        return m_node.mempool->exists(GenTxid::Txid(txid));
+        return m_node.mempool->exists(Txid::FromUint256(txid));
```
Good looks

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32631/commits/a60f863d3e276534444571282f432b913d3967db

dergoegge reviewed (approved):
Code review ACK a60f863d3e276534444571282f432b913d3967db

maflcko reviewed (approved):
left two nits that can be ignored or followed-up  later on.

review ACK a60f863d3e276534444571282f432b913d3967db 🎽

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK a60f863d3e276534444571282f432b913d3967db 🎽
Y4M+5Nl4FOcgYfClv1qRnd5797LuKUVoDluobop5CMpD3562zy81q/G5Wgo0lri3tMAlaUzCOjdjwFe/zausCw==
```

</details>

theStack reviewed (approved):
Code-review ACK a60f863d3e276534444571282f432b913d3967db

JeremyRubin commented:
i get that this is an overall improvement to the type safety, but I think that it's gone about sort of the wrong way since it leaves a lot of data-structures that are supposed to be made homogeneous as hetergeneous still. E.g.,         

```
std::set<GenTxid> m_tx_inventory_to_send;
 ```

is _supposed_ to be     `std::variant<std::set<Txid>, std::set<WTxid>>`.

You can create a templated wrapper to encapsulate a generic container interface for this.

hodlinator reviewed (commented):
Light post-merge ACK

marcofleon commented:
Follow ups to these last few comments can be found here: https://github.com/bitcoin/bitcoin/pull/33005
